testplan/runners/pools/remote.py

```diff
@@ -385,6 +398,47 @@ def _start_thread_pool(self) -> None:
                     "Please upgrade to the suggested python interpreter."
                 )
 
+    def _stop_worker_loop(self):
+        while self.active:
+            worker_id = None
+            with self._pool_lock:
+                try:
+                    worker_id = self._wait_stop_queue.pop(0)
+                    self._stopping_queue.append(worker_id)
+                except IndexError:
+                    pass
+
+            if worker_id:
+                self.logger.debug(
+                    "Get early stop worker %s in queue", worker_id
+                )
+                worker = self._workers[worker_id]
+                if self.pool:
+                    self.pool.apply_async(
+                        self._workers.sync_stop_resource, (worker,)
+                    )
+                else:
+                    self._workers.sync_stop_resource(worker)
+            else:
+                time.sleep(self.QUEUE_WAIT_INTERVAL)
+
+    def _early_stop_worker(self, worker_uid: Union[str, int]) -> bool:
+        with self._pool_lock:
+            if (
+                worker_uid not in self._wait_stop_queue
+                and worker_uid not in self._stopping_queue
+            ):
+                if self.RESERVE_WORKER_NUM + self.unassigned.size() < len(
+                    self.worker_status["active"]
+                ) - len(self._wait_stop_queue) - len(self._stopping_queue):
+                    self.logger.user_info(
+                        "Early stop worker %s", self._workers[worker_uid]
+                    )
+                    print(f"Early stop worker {self._workers[worker_uid]}")
+                    self._wait_stop_queue.append(worker_uid)
```

Comment:
I think we can do self.pool.apply_sync directly, and save that _stop_worker_thread.